File: ios/JustSpent/JustSpent/ViewModels/ExpenseListViewModel.swift
import Foundation
import Combine
import CoreData

// MARK: - Pagination State
struct PaginationState {
    let loadedExpenses: [Expense]
    let currentPage: Int
    let hasMore: Bool
    let isLoading: Bool
    let error: String?

    init(
        loadedExpenses: [Expense] = [],
        currentPage: Int = 0,
        hasMore: Bool = false,
        isLoading: Bool = false,
        error: String? = nil
    ) {
        self.loadedExpenses = loadedExpenses
        self.currentPage = currentPage
        self.hasMore = hasMore
        self.isLoading = isLoading
        self.error = error
    }
}

@MainActor
class ExpenseListViewModel: ObservableObject {
    @Published var expenses: [Expense] = []
    @Published var isLoading = false
    @Published var errorMessage: String?

    // Pagination state
    @Published var paginationState = PaginationState()

    private let repository: ExpenseRepositoryProtocol
    private var cancellables = Set<AnyCancellable>()

    // Pagination properties
    private var currentCurrency: String = ""
    private var currentDateFilter: DateFilter = .all
    private let pageSize: Int = 20

    init(repository: ExpenseRepositoryProtocol = ExpenseRepository(), context: NSManagedObjectContext? = nil) {
        self.repository = repository
        loadExpenses()
    }
    
    func loadExpenses() {
        isLoading = true
        errorMessage = nil
        
        repository.getAllExpenses()
            .receive(on: DispatchQueue.main)
            .sink(
                receiveCompletion: { [weak self] completion in
                    self?.isLoading = false
                    if case .failure(let error) = completion {
                        self?.errorMessage = error.localizedDescription
                    }
                },
                receiveValue: { [weak self] expenses in
                    self?.expenses = expenses
                }
            )
            .store(in: &cancellables)
    }
    
    func addSampleExpense() async {
        do {
            let expenseData = ExpenseData(
                amount: NSDecimalNumber(value: Double.random(in: 5.0...50.0)),
                currency: "USD",
                category: ["Food & Dining", "Grocery", "Transport", "Shopping"].randomElement()!,
                merchant: ["Coffee Shop", "Supermarket", "Gas Station", "Store"].randomElement()!,
                notes: "Sample expense",
                transactionDate: Date(),
                source: "manual",
                voiceTranscript: nil
            )
            
            _ = try await repository.addExpense(expenseData)
            
        } catch {
            await MainActor.run {
                self.errorMessage = error.localizedDescription
            }
        }
    }
    
    func deleteExpense(_ expense: Expense) async {
        do {
            try await repository.deleteExpense(expense)
        } catch {
            await MainActor.run {
                self.errorMessage = error.localizedDescription
            }
        }
    }
    
    var totalSpending: Double {
        expenses.reduce(0) { total, expense in
            total + (expense.amount?.doubleValue ?? 0)
        }
    }
    
    var formattedTotalSpending: String {
        let formatter = NumberFormatter()
        formatter.numberStyle = .currency
        formatter.currencyCode = "USD"
        return formatter.string(from: NSNumber(value: totalSpending)) ?? "$0.00"
    }

    // MARK: - Pagination Methods
    func loadFirstPage(currency: String, dateFilter: DateFilter) async {
        currentCurrency = currency
        currentDateFilter = dateFilter

        paginationState = PaginationState(isLoading: true, error: nil)

        do {
            let expenses = try await repository.loadExpensesPage(
                currency: currency,
                dateFilter: dateFilter,
                page: 0,
                pageSize: pageSize
            )

            paginationState = PaginationState(
                loadedExpenses: expenses,
                currentPage: 0,
                hasMore: expenses.count == pageSize,
                isLoading: false,
                error: nil
            )
        } catch {
            paginationState = PaginationState(
                loadedExpenses: [],
                currentPage: 0,
                hasMore: false,
                isLoading: false,
                error: error.localizedDescription
            )
        }
    }

    func loadNextPage() async {
        // Don't load if already loading or no more data
        guard !paginationState.isLoading, paginationState.hasMore else {
            return
        }

        let nextPage = paginationState.currentPage + 1

        paginationState = PaginationState(
            loadedExpenses: paginationState.loadedExpenses,
            currentPage: paginationState.currentPage,
            hasMore: paginationState.hasMore,
            isLoading: true,
            error: nil
        )

        do {
            let newExpenses = try await repository.loadExpensesPage(
                currency: currentCurrency,
                dateFilter: currentDateFilter,
                page: nextPage,
                pageSize: pageSize
            )

            paginationState = PaginationState(
                loadedExpenses: paginationState.loadedExpenses + newExpenses,
                currentPage: nextPage,
                hasMore: newExpenses.count == pageSize,
                isLoading: false,
                error: nil
            )
        } catch {
            paginationState = PaginationState(
                loadedExpenses: paginationState.loadedExpenses,
                currentPage: paginationState.currentPage,
                hasMore: paginationState.hasMore,
                isLoading: false,
                error: error.localizedDescription
            )
        }
    }
}

File: android/app/src/main/java/com/justspent/expense/ui/expenses/ExpenseListViewModel.kt
package com.justspent.expense.ui.expenses

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.justspent.expense.data.model.Currency
import com.justspent.expense.data.model.Expense
import com.justspent.expense.data.model.ExpenseData
import com.justspent.expense.data.preferences.UserPreferences
import com.justspent.expense.data.repository.ExpenseRepositoryInterface
import com.justspent.expense.utils.CurrencyFormatter
import com.justspent.expense.utils.DateFilter
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch
import kotlinx.datetime.Clock
import kotlinx.datetime.TimeZone
import kotlinx.datetime.toLocalDateTime
import java.math.BigDecimal
import javax.inject.Inject

data class ExpenseListUiState(
    val expenses: List<Expense> = emptyList(),
    val isLoading: Boolean = false,
    val errorMessage: String? = null,
    val totalSpending: BigDecimal = BigDecimal.ZERO,
    val formattedTotalSpending: String = "$0.00"
)

data class PaginationState(
    val loadedExpenses: List<Expense> = emptyList(),
    val currentPage: Int = 0,
    val hasMore: Boolean = false,
    val isLoading: Boolean = false,
    val error: String? = null
)

@HiltViewModel
class ExpenseListViewModel @Inject constructor(
    private val repository: ExpenseRepositoryInterface,
    private val userPreferences: UserPreferences
) : ViewModel() {

    private val _uiState = MutableStateFlow(ExpenseListUiState())
    val uiState: StateFlow<ExpenseListUiState> = _uiState.asStateFlow()

    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage.asStateFlow()

    // Pagination state
    private val _paginationState = MutableStateFlow(PaginationState())
    val paginationState: StateFlow<PaginationState> = _paginationState.asStateFlow()

    private var currentCurrency: String = ""
    private var currentDateFilter: DateFilter = DateFilter.All
    private val pageSize: Int = 20

    init {
        loadExpenses()
    }

    private fun loadExpenses() {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true, errorMessage = null) }

            try {
                combine(
                    repository.getAllExpenses(),
                    repository.getTotalSpending(),
                    userPreferences.defaultCurrency
                ) { expenses, total, defaultCurrency ->
                    val totalAmount = total ?: BigDecimal.ZERO
                    val formattedTotal = formatCurrency(totalAmount, defaultCurrency)

                    ExpenseListUiState(
                        expenses = expenses,
                        isLoading = false,
                        totalSpending = totalAmount,
                        formattedTotalSpending = formattedTotal
                    )
                }.collect { state ->
                    _uiState.value = state
                }
            } catch (e: Exception) {
                _uiState.update {
                    it.copy(
                        isLoading = false,
                        errorMessage = e.message ?: "Unknown error"
                    )
                }
            }
        }
    }

    fun addSampleExpense() {
        viewModelScope.launch {
            val categories = listOf("Food & Dining", "Grocery", "Transport", "Shopping")
            val merchants = listOf("Coffee Shop", "Supermarket", "Gas Station", "Store")
            val amounts = listOf(5.50, 15.75, 25.00, 45.99, 8.25)

            val expenseData = ExpenseData(
                amount = BigDecimal(amounts.random()),
                currency = "USD",
                category = categories.random(),
                merchant = merchants.random(),
                notes = "Sample expense",
                transactionDate = Clock.System.now().toLocalDateTime(TimeZone.currentSystemDefault()),
                source = "manual"
            )

            repository.addExpense(expenseData)
                .onFailure { error ->
                    _errorMessage.value = error.message
                }
        }
    }

    fun deleteExpense(expense: Expense) {
        viewModelScope.launch {
            repository.deleteExpense(expense)
                .onFailure { error ->
                    _errorMessage.value = error.message
                }
        }
    }

    fun updateExpense(expense: Expense) {
        viewModelScope.launch {
            repository.updateExpense(expense)
                .onFailure { error ->
                    _errorMessage.value = error.message
                }
        }
    }

    fun clearErrorMessage() {
        _errorMessage.value = null
    }

    private fun formatCurrency(amount: BigDecimal, currency: Currency): String {
        return CurrencyFormatter.format(
            amount = amount,
            currency = currency,
            showSymbol = true,
            showCode = false
        )
    }

    // Pagination methods
    fun loadFirstPage(currency: String, dateFilter: DateFilter) {
        currentCurrency = currency
        currentDateFilter = dateFilter

        viewModelScope.launch {
            _paginationState.update { it.copy(isLoading = true, error = null) }

            repository.loadExpensesPage(
                currency = currency,
                dateFilter = dateFilter,
                page = 0,
                pageSize = pageSize
            ).fold(
                onSuccess = { expenses ->
                    _paginationState.update {
                        PaginationState(
                            loadedExpenses = expenses,
                            currentPage = 0,
                            hasMore = expenses.size == pageSize,
                            isLoading = false,
                            error = null
                        )
                    }
                },
                onFailure = { error ->
                    _paginationState.update {
                        it.copy(
                            isLoading = false,
                            error = error.message ?: "Failed to load expenses"
                        )
                    }
                }
            )
        }
    }

    fun loadNextPage() {
        val currentState = _paginationState.value

        // Don't load if already loading or no more data
        if (currentState.isLoading || !currentState.hasMore) {
            return
        }

        val nextPage = currentState.currentPage + 1

        viewModelScope.launch {
            _paginationState.update { it.copy(isLoading = true, error = null) }

            repository.loadExpensesPage(
                currency = currentCurrency,
                dateFilter = currentDateFilter,
                page = nextPage,
                pageSize = pageSize
            ).fold(
                onSuccess = { newExpenses ->
                    _paginationState.update { state ->
                        state.copy(
                            loadedExpenses = state.loadedExpenses + newExpenses,
                            currentPage = nextPage,
                            hasMore = newExpenses.size == pageSize,
                            isLoading = false,
                            error = null
                        )
                    }
                },
                onFailure = { error ->
                    _paginationState.update {
                        it.copy(
                            isLoading = false,
                            error = error.message ?: "Failed to load more expenses"
                        )
                    }
                }
            )
        }
    }
}